Bug fix for behavior after exceptions with concurrent lumis

wddgit opened the pull request:
#### PR description:

This fixes a bug that would be encountered when more than
one lumi is being processed concurrently and there is an
exception. If the exception is associated with a lumi other
than the last one read, an infinite wait will be entered.
The wait is in the processLumis function of the EventProcessor.
It waits on every task in the stream SerialTaskQueues.
In the above situation the serial queue is not resumed
before the wait and there are still tasks in the queues.
There is code that would clean these up in the function
endUnfinishedLumis but that only runs after the wait
is over and processLumis has returned. Too late.

This commit also fixes the maxEvents output parameter which was
broken when more than one lumi is processed concurrently. As far
as I know nothing uses this, but it is an advertised parameter
of the Framework that is supposed to work. One note about
this. If events are running concurrently when the limit is
reached, then all the events already running will complete, so
the job could write more than the requested number of events
to output.

The modified code is not executed if there are no exceptions and
the maxEvents output parameter is not used. Under normal
circumstances this should not have any effect on output or
performance.

#### PR validation:

This adds two new unit tests. One test would fail with an infinite
wait if executed in a release before this commit. It creates the
condition described above. The other tests the maxEvent output
parameter. Plus I added an assert that checks that the reference
count is one in normalEnd when we try to delete RunResources
and get endRun to execute (something to be concerned about in
concurrent lumi mode).

wddgit committed:
Bug fix for behavior after exceptions with concurrent lumis

This fixes a bug that would be encountered when more than
one lumi is being processed concurrently and there is an
exception. If the exception is associated with a lumi other
than the last one read, an infinite wait will be entered.
The wait is in the processLumis function of the EventProcessor.
It waits on every task in the stream SerialTaskQueues.
In the above situation the serial queue is not resumed
before the wait and there are still tasks in the queues.
There is code that would clean these up in the function
endUnfinishedLumis but that only runs after the wait
is over and processLumis has returned. Too late.

This commit also fixes the maxEvent output parameter which was
broken when more than one lumi is processed concurrently. As far
as I know nothing uses this, but it is an advertised parameter
of the Framework that is supposed to work. One note about
this. If events are running concurrently when the limit is
reached, then all the events already running will complete, so
the job could write more than the requested number of events
to output.

This adds two new unit tests. One test would fail with an infinite
wait if executed in a release before this commit. The other tests
the maxEvent output parameter. Plus I added an assert that
checks that the reference count is one in normalEnd when we
try to delete RunResources and get endRun to execute.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26349/9064

- This PR adds an extra 48KB to repository

cmsbuild commented:
A new Pull Request was created by  @wddgit (W. David Dagenhart) for master.

It involves the following packages:

FWCore/Framework


@cmsbuild, @smuzaffar, @Dr15Jones can you please review it and eventually sign? Thanks.
@makortel, @Martin-Grunewald this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

wddgit commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 1259b5bb3e32e03881c776e8b019d45b9acc0895

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26349/33947/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test TestFWCoreFrameworkCmsRun had ERRORS

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26349/33947/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3139747
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3139549
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

Dr15Jones commented:
It looks like the new test failed?

```
Failure running cmsRun /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_6_X_2019-04-03-1100/src/FWCore/Framework/test/testMaxEventsOutput_cfg.py: status 65
```

wddgit committed:
Fix previous commit

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26349/9079

- This PR adds an extra 52KB to repository

cmsbuild commented:
Pull request #26349 was updated. @cmsbuild, @smuzaffar, @Dr15Jones can you please check and sign again.

wddgit commented:
please test

I had the unit tests working, then at the last minute I experimented with some minor test changes to check something and failed to correctly restore it to what I originally intended. Try again.

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 739dd4534a57364d6159076aac91cdf2b4421e01
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26349/33964/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26349/33964/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3139747
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3139549
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

Dr15Jones commented on FWCore/Framework/src/EventProcessor.cc:
```diff
@@ -1155,7 +1158,9 @@ namespace edm {
                   {
                     if(iPtr) {
-                      h.doneWaiting(*iPtr);
+                      WaitingTaskHolder tmp(h);
```
For my own notes: making the copy avoids having `dineWaiting` launch the task before the call to `streamEndLumiAsync`.

Dr15Jones commented on FWCore/Framework/src/EventProcessor.cc:
```diff
@@ -1747,7 +1788,10 @@ namespace edm {
   void EventProcessor::setExceptionMessageLumis(std::string& message) {
-    exceptionMessageLumis_ = message;
+    bool expected = false;
+    if (exceptionMessageLumisIsSet_.compare_exchange_strong(expected,true)) {
```
This is not technically thread safe. The `exceptionMessageLumisIsSet_` is not functioning as a proper guard since it does not guarantee that after writing to `exceptionMessageLumis_` that memory is synchronized across all threads that might read it. In practice, when the `tbb::task` that called `setExceptionMessageLumis` completes, it will synchronize all memory written during the task's exectution.

The better way to do this is to mae `exceptionMessageLumis_ ` a `std::atomic<std::string*>` and handle memory cleanup at the point which looks at `exceptionMessageLumis_`.

wddgit replied:
Agreed. Although nothing was actually reading it in the concurrent parts of the code, but I agree it is a landmine we might possibly hit later. I thought of a simpler fix. The string is always the same now.  It is simpler just to delete the string data member entirely. It is not needed. All we need is an atomic<bool> data member. Fixed in the commit I just pushed.

wddgit committed:
Improve exception handling for end stream exceptions

These improvements are particularly relevant when
running with concurrent lumis.

More thread safe. Just use an atomic<bool> for exceptionMessageLumis_.
No reason to pass around the string. It's always the same.

Set the deferredExceptionPtr_ when there are exceptions in end lumi
transitions to shut down quicker and with less noise in the printed
exceptions.

Some cleanup to slightly simplify this part of the code.

cmsbuild commented:
The code-checks are being triggered in jenkins.

wddgit commented:
I just pushed a fix to Chris's push. And as I looked at it some more I saw some other ways to improve it. It is simplified in a few places and also will behave better when there are multiple exceptions getting raised, usually a primary one and others that are side effects of the primary exception.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26349/9117

- This PR adds an extra 56KB to repository

wddgit commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #26349 was updated. @cmsbuild, @smuzaffar, @Dr15Jones can you please check and sign again.

cmsbuild commented:
+1
Tested at: 2d43723df62bc5920eaa633a9071fbf7d7ada410
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26349/34021/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26349/34021/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3140495
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3140297
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

Dr15Jones commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
+1
